### DISCUSSION TITLE
Argument of type '"query"' is not assignable to parameter of type 'never'.ts(2345) on prisma.$on()

### DISCUSSION THREAD

**Discussion by aflatoon2874** [prisma/prisma#3541] [Q&A] [answered]
I am getting the above error while defining an event handler. Example:

```ts
prisma.$on('query', async (e) => {
  await fs.promises.appendFile(
    __dirname + "/test.sql",
    `${e.query} ${e.params} \n`
  );
});
```
The editor intellisense shows the method signature as:
(method) PrismaClient<PrismaClientOptions, never>.$on<never>(eventType: never, callback: (event: never) => void): void

How to handle this type error? Please help. Thanks

**Comment by pantharshit00** [comment]
@aflatoon2874 

You will need to enable `"emit": "event"` in the prisma constructor if you want to use this feature:
```js
const prisma = new PrismaClient({
  log: [
    {
      emit: "event",
      level: "query",
    },
  ],
});
```

**Reply by aflatoon2874** [reply]
@pantharshit00 
I did that. I did not mention as it was understood. Why the signature is so weird? How can arguments be typed "never"? I have tried with v2.5.1 and v2.6.2. It does not work.

**Reply by aflatoon2874** [reply]
Ok. on further investigation, I found out that in one scenario it works but in other one it fails and throws the above cited error.

Scenario 1:  This is how I was doing it and getting the error
```
const prismaOptions: PrismaClientOptions = {
	log: [
		{
			emit: 'stdout',
			level: 'info'
		},
		{
			emit: 'stdout',
			level: 'warn'
		},
		{
			emit: 'event',
			level: 'query'
		}
	]
}

const prisma = new PrismaClient(prismaOptions)

// define prisma event handlers
prisma.$on('query', async (e) => {
  await fs.promises.appendFile(
    `${__dirname}/test.sql`,
    `${e.query} ${e.params} \n`
  );
});
```
Scenario 2: Directly passing the object which works
```
const prisma = new PrismaClient({
	log: [
		{
			emit: 'stdout',
			level: 'info'
		},
		{
			emit: 'stdout',
			level: 'warn'
		},
		{
			emit: 'event',
			level: 'query'
		}
	]
})

// define prisma event handlers
prisma.$on('query', async (e) => {
  await fs.promises.appendFile(
    `${__dirname}/test.sql`,
    `${e.query} ${e.params} \n`
  );
});
```
Can you please shed some light on this strange behaviour. In my opinion, both should work.

**Reply by pantharshit00** [reply]
@aflatoon2874 If you type the option object we can't validate on a type level as they will allow every value to be passed.

The explanation is kind of same as we have here: https://github.com/prisma/prisma/issues/3372#issuecomment-688294309

**Reply by aflatoon2874** [reply]
Thanks @pantharshit00 .
How do I programmatically set the sql query capture for non-production environments. I want to configure something like
```
if (process.env.NODE_ENV !== 'production') {
	prismaOptions?.log?.push({
		emit: 'event',
		level: 'query'
	})
}
```

**Reply by yanavery** [reply]
I know this is an old thread, sorry to revive, but how was this resolved, if at all? Thanks for sharing your solution/workaround ...

**Reply by aflatoon2874** [reply, marked as answer]
> I know this is an old thread, sorry to revive, but how was this resolved, if at all? Thanks for sharing your solution/workaround ...

To be frank I did not spend any more time and went ahead with the workaround. I am reproducing verbatim my implementation of the `getPrismaClient` function. It supports multi-tenant databases having the same schema. Hope this helps.

```
import { PrismaClient } from '@prisma/client'
import fs from 'fs'

export const getPrismaClient = (dbUrl?: string): PrismaClient => {
  if (!dbUrl) {
    dbUrl = 'default'
  }
  if (process.env.NODE_ENV === 'production') {
    return getPrismaClientForProduction(dbUrl)
  } else {
    return getPrismaClientForDevelopment(dbUrl)
  }
}

const getPrismaClientForProduction = (dbUrl: string) => {
  if (dbUrl === 'default') {
    return new PrismaClient({
      log: [
        {
          emit: 'stdout',
          level: 'info'
        },
        {
          emit: 'stdout',
          level: 'warn'
        }
      ]
    })
  } else {
    return new PrismaClient({
      log: [
        {
          emit: 'stdout',
          level: 'info'
        },
        {
          emit: 'stdout',
          level: 'warn'
        }
      ],
      datasources: {
        db: {
          url: dbUrl
        }
      }
    })
  } // end if
}

const getPrismaClientForDevelopment = (dbUrl: string) => {
  let prisma
  if (dbUrl === 'default') {
    prisma = new PrismaClient({
      log: [
        {
          emit: 'stdout',
          level: 'info'
        },
        {
          emit: 'stdout',
          level: 'warn'
        },
        {
          emit: 'event',
          level: 'query'
        }
      ]
    })
  } else {
    prisma = new PrismaClient({
      log: [
        {
          emit: 'stdout',
          level: 'info'
        },
        {
          emit: 'stdout',
          level: 'warn'
        },
        {
          emit: 'event',
          level: 'query'
        }
      ],
      datasources: {
        db: {
          url: dbUrl
        }
      }
    })
  } // end if

  // define prisma event handlers
  prisma.$on('query', async (e) => {
    await fs.promises.appendFile(`generated.sql`, `${e.query} ${e.params} \n`)
  })

  return prisma
}
```

**Comment by JavierGMo** [comment]
prisma.$on("query" as never, async (e: Prisma.QueryEvent) => {
      console.log(`${e.query} ${e.params}`);
    });

I found in stack overflow https://stackoverflow.com/questions/73654056/how-to-see-prisma-query-values

**Reply by mochadwi** [reply]
This was works for me at least, to ignore the "type error and such"

**Comment by kasir-barati** [comment]
### JFYI if you're having trouble in NestJS check how I did it here:

https://github.com/kasir-barati/graphql/blob/f42dbd02cb61d82b782f2107859b0472bae4afa1/libs/shared/src/modules/prisma/prisma.service.ts#L1-L40

**Reply by mochadwi** [reply]
Thanks for pointing out this one, I'm also using NestJS
